pipeline-runner/R/seurat-3-upload_seurat_to_aws.R

```diff
@@ -64,6 +77,69 @@ upload_seurat_to_aws <- function(input, pipeline_config, prev_out) {
   return(res)
 }
 
+find_cluster_columns <- function(scdata) {
+
+  # exclude all group columns, including duplicates
+  group_cols <- find_group_columns(scdata, remove.dups = FALSE)
+  exclude_cols <- c(group_cols, 'samples')
+
+  # order meta to indicate preference for louvain clusters
+  meta <- scdata@meta.data
+  louvain_cols <- c('active.ident', 'seurat_clusters')
+  meta <- meta |> dplyr::relocate(dplyr::any_of(louvain_cols))
+
+  check_cols <- setdiff(colnames(meta), exclude_cols)
+
+  cluster_cols <- c()
+  for (check_col in check_cols) {
+    check_vals <- meta[[check_col]]
+
+    # skip if too few or way too many values
+    n.vals <- length(table(check_vals))
+    if (n.vals < 2) next()
+    if (n.vals > 1000) next()
+    cat("candidate cluster columns:", check_col, "--> nvals:", n.vals, '\n')
+
+    # skip if values are numeric but non-integer
+    if (is.numeric(check_vals) &&
+        !all(as.integer(check_vals) == check_vals)) next()
+
+    # skip if more than 1/3 of values are repeated fewer than 4 times
+    value_counts <- table(check_vals)
+    nreps_lt4 <- sum(value_counts < 4)
+    if (nreps_lt4 > length(unique(check_vals))/3) next()
+
+    # skip if col is same as samples or group column
+    is_sample_col <- FALSE
+    for (exclude_col in exclude_cols)  {
+      exclude_vals <- meta[[exclude_col]]
+      if (test_groups_equal(check_vals, exclude_vals)) {
+        is_sample_col <- TRUE
+        break
+      }
+    }
+    if (is_sample_col) next()
+
+    # TODO: decide if should skip if values are boolean?
```

Comment:
I'm curious about this. what would be the usecase?

Reply:
Just that a column could have boolean values e.g. `isDoublet` or `isEpithelial`. We don't necessarily want to include the former as a cellset but want to include the later. It's really just a particular way of encoding a two-valued cellset division (which would pass all other checks) so I elected not to exclude them for the initial implementation.

Reply:
yeah, one-hot encoding features makes all of this more complicated. It's something that could be required to address by the users (I'm also considering this case for the cell-level metadata feature)